# Case Study Project - Airbnb Berlin
# Table of Contents
## 1. Import libraries
## 2. Import Data
## 3. Data Wrangling
## 4. Data Cleaning and Consistency Check
- check for missing data
- check for duplicates
- check for mixed data types 

## 5. Export Data

# 01. Import Libraries

In [23]:
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [24]:
# creating a path
path = r'/Users/sudip/Desktop/DATA ANALYST/Projects/Project 6/Data'

In [25]:
# importing dataset
df = pd.read_csv(os.path.join(path, 'Original Data', 'airbnb_berlin', 'listings.csv'), index_col = False)

# 03 Data Wrangling

In [26]:
# Checking shape
df.shape

(16416, 18)

In [27]:
# checking data types
df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                            object
dtype: object

In [28]:
# checking the information of each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16416 entries, 0 to 16415
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              16416 non-null  int64  
 1   name                            16392 non-null  object 
 2   host_id                         16416 non-null  int64  
 3   host_name                       16406 non-null  object 
 4   neighbourhood_group             16416 non-null  object 
 5   neighbourhood                   16416 non-null  object 
 6   latitude                        16416 non-null  float64
 7   longitude                       16416 non-null  float64
 8   room_type                       16416 non-null  object 
 9   price                           16416 non-null  int64  
 10  minimum_nights                  16416 non-null  int64  
 11  number_of_reviews               16416 non-null  int64  
 12  last_review                     

In [29]:
df.describe()

id       host_id      latitude     longitude         price  \
count  1.641600e+04  1.641600e+04  16416.000000  16416.000000  16416.000000   
mean   2.613804e+16  1.006091e+08     52.509612     13.404221     85.764559   
std    1.220472e+17  1.184267e+08      0.035503      0.069988    104.924635   
min    3.176000e+03  1.581000e+03     52.321950     13.072850      0.000000   
25%    1.270874e+07  1.241388e+07     52.489317     13.367260     39.000000   
50%    2.481971e+07  4.537750e+07     52.509770     13.414405     60.000000   
75%    4.094544e+07  1.543070e+08     52.533140     13.439312    100.000000   
max    6.471177e+17  4.633970e+08     52.673960     13.769390   4060.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    16416.000000       16416.000000       13533.000000   
mean        11.463024          26.155032           0.770905   
std         37.378188          59.065977           1.612770   
min          1.000000           0.000000           0.010000   
25%          2.000000           1.000000           0.080000   
50%          3.000000           5.000000           0.260000   
75%          5.000000          21.000000           0.900000   
max       1124.000000        1269.000000          86.910000   

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
count                    16416.000000      16416.000000           16416.000000  
mean                         3.151803         78.865192               5.763523  
std                          6.719214        115.215045              19.405979  
min                          1.000000          0.000000               0.000000  
25%                          1.000000          0.000000               0.000000  
50%                          1.000000          0.000000               0.000000  
75%                          2.000000        158.000000               4.000000  
max                         50.000000        365.000000            1049.000000

In [30]:
df.head()

id                                               name  \
0            39971294                             Cozy flat with chimney   
1            11934470        Bright Studio Apartment / 4th fl/ long term   
2            17580693  Gemütliches, Zentrales Zimmer mit Blick ins Grüne   
3  579471587567149993               Süßes Altbau-Apartment in Schöneberg   
4            17497917            BERLIN MITTE - hotspot - central & calm   

     host_id       host_name       neighbourhood_group          neighbourhood  \
0   21565256           Tanja  Friedrichshain-Kreuzberg  nördliche Luisenstadt   
1   62819516  Carla & Daniel                  Neukölln           Reuterstraße   
2  119416956            Sven                     Mitte            Parkviertel   
3    9916942         Kirsten    Tempelhof - Schöneberg              Friedenau   
4    5937313       Hyun-Jung                     Mitte         Alexanderplatz   

    latitude  longitude        room_type  price  minimum_nights  \
0  52.501680  13.415730  Entire home/apt     50               1   
1  52.492260  13.429370  Entire home/apt     43              30   
2  52.561510  13.355150     Private room     42               2   
3  52.477266  13.336424  Entire home/apt     52              92   
4  52.524380  13.410120  Entire home/apt    106               2   

   number_of_reviews last_review  reviews_per_month  \
0                  2  2019-12-02               0.06   
1                  5  2022-03-29               0.07   
2                 31  2022-03-07               0.50   
3                  0         NaN                NaN   
4                116  2022-05-29               1.84   

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  \
0                               1                 0                      0   
1                               2               271                      3   
2                               1                10                      4   
3                               1               362                      0   
4                               1                 6                     25   

  license  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

### After initially exploring the dataset, we can make several observations.  There are 18 columns and 16,416 observations. There are a few missing values, as well as many NaN observations under the license column. There are missing values in name,  host_name , last_review and reviews_per_month columnc as well. 

In [31]:
df.isnull().sum()

id                                    0
name                                 24
host_id                               0
host_name                            10
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                        2883
reviews_per_month                  2883
calculated_host_listings_count        0
availability_365                      0
number_of_reviews_ltm                 0
license                           11433
dtype: int64

### Based on above results, there are several missing observations in 'name' and 'host_name'. However, due to PII data concerns, host_name will be removed from the dataset.
### Furthermore, there are 2883 columns with missing values for last_review and reviews_per_month columns. As we do not have further information about 'Nan' values in last_review, we will delete  rows with NaN values in last_review.
### We will delete licence columns as this column is irrelevant for this analysis.

# 04. Data Cleaning

## Dropping Columns/Variables

In [32]:
# Dropping columns that contain PII data and are irrelavant for this analysis: 
df.drop(['host_name', 'license'], axis=1, inplace=True)
df.head()

id                                               name  \
0            39971294                             Cozy flat with chimney   
1            11934470        Bright Studio Apartment / 4th fl/ long term   
2            17580693  Gemütliches, Zentrales Zimmer mit Blick ins Grüne   
3  579471587567149993               Süßes Altbau-Apartment in Schöneberg   
4            17497917            BERLIN MITTE - hotspot - central & calm   

     host_id       neighbourhood_group          neighbourhood   latitude  \
0   21565256  Friedrichshain-Kreuzberg  nördliche Luisenstadt  52.501680   
1   62819516                  Neukölln           Reuterstraße  52.492260   
2  119416956                     Mitte            Parkviertel  52.561510   
3    9916942    Tempelhof - Schöneberg              Friedenau  52.477266   
4    5937313                     Mitte         Alexanderplatz  52.524380   

   longitude        room_type  price  minimum_nights  number_of_reviews  \
0  13.415730  Entire home/apt     50               1                  2   
1  13.429370  Entire home/apt     43              30                  5   
2  13.355150     Private room     42               2                 31   
3  13.336424  Entire home/apt     52              92                  0   
4  13.410120  Entire home/apt    106               2                116   

  last_review  reviews_per_month  calculated_host_listings_count  \
0  2019-12-02               0.06                               1   
1  2022-03-29               0.07                               2   
2  2022-03-07               0.50                               1   
3         NaN                NaN                               1   
4  2022-05-29               1.84                               1   

   availability_365  number_of_reviews_ltm  
0                 0                      0  
1               271                      3  
2                10                      4  
3               362                      0  
4                 6                     25

In [33]:
#checking for null/missing values
df.isna().sum()

id                                   0
name                                24
host_id                              0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       2883
reviews_per_month                 2883
calculated_host_listings_count       0
availability_365                     0
number_of_reviews_ltm                0
dtype: int64

## Replacing missing values

In [34]:
# As we donot have any information about 'NaN' values in last_review column, dropping the index with last_review  = Nan
df.dropna(subset = ["last_review"], inplace=True)

In [35]:
#checking for null/missing values
df.isna().sum()

id                                 0
name                              12
host_id                            0
neighbourhood_group                0
neighbourhood                      0
latitude                           0
longitude                          0
room_type                          0
price                              0
minimum_nights                     0
number_of_reviews                  0
last_review                        0
reviews_per_month                  0
calculated_host_listings_count     0
availability_365                   0
number_of_reviews_ltm              0
dtype: int64

In [37]:
# Replacing all NaN values under "name" with 'Unknown'
df.fillna({'name':'Unknown'}, inplace=True)

In [38]:
#checking for null/missing values
df.isna().sum()

id                                0
name                              0
host_id                           0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
dtype: int64

In [39]:
df.shape

(13533, 16)

## Consistency Check: Categorical Variables

In [40]:
df['room_type'].value_counts()

Entire home/apt    8030
Private room       5232
Shared room         151
Hotel room          120
Name: room_type, dtype: int64

In [41]:
df['neighbourhood_group'].value_counts()

Friedrichshain-Kreuzberg    3004
Mitte                       2909
Pankow                      2180
Neukölln                    1835
Charlottenburg-Wilm.        1014
Tempelhof - Schöneberg       910
Treptow - Köpenick           475
Lichtenberg                  407
Steglitz - Zehlendorf        358
Reinickendorf                192
Spandau                      140
Marzahn - Hellersdorf        109
Name: neighbourhood_group, dtype: int64

In [42]:
# checking for duplicates
df_dups = df[df.duplicated()]
df_dups

Empty DataFrame
Columns: [id, name, host_id, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365, number_of_reviews_ltm]
Index: []

## There are no dublicates in the dataframe.

In [43]:
# checking for mixed data type
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

## There no mixed data types.

In [44]:
# Checking the data shape of cleaned data.
df.shape

(13533, 16)

In [45]:
df.describe()

id       host_id      latitude     longitude         price  \
count  1.353300e+04  1.353300e+04  13533.000000  13533.000000  13533.000000   
mean   1.569264e+16  9.567787e+07     52.509526     13.404592     84.987955   
std    9.477258e+16  1.140730e+08      0.035171      0.069812     94.375962   
min    3.176000e+03  1.581000e+03     52.332300     13.072850      0.000000   
25%    1.240587e+07  1.085170e+07     52.489380     13.368900     40.000000   
50%    2.378688e+07  4.269422e+07     52.509770     13.414940     60.000000   
75%    3.947761e+07  1.464621e+08     52.533100     13.439280     99.000000   
max    6.432860e+17  4.607175e+08     52.673960     13.769390   4060.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    13533.000000       13533.000000       13533.000000   
mean         9.845858          31.726964           0.770905   
std         35.794265          63.681117           1.612770   
min          1.000000           1.000000           0.010000   
25%          2.000000           3.000000           0.080000   
50%          3.000000           8.000000           0.260000   
75%          5.000000          28.000000           0.900000   
max       1124.000000        1269.000000          86.910000   

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
count                    13533.000000      13533.000000           13533.000000  
mean                         2.996379         79.161901               6.991354  
std                          6.192658        113.889181              21.171687  
min                          1.000000          0.000000               0.000000  
25%                          1.000000          0.000000               0.000000  
50%                          1.000000          0.000000               1.000000  
75%                          2.000000        158.000000               6.000000  
max                         50.000000        365.000000            1049.000000

# 05. Exporting Data

In [46]:
# exporting the clean data
df.to_csv(os.path.join(path,'prepared_data','airbnb_berlin_cleaned.csv'))